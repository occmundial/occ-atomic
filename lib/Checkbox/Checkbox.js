import React from 'react';
import PropTypes from 'prop-types';

import Text from '../Text';
import Icon from '../Icon';
import colors from '../tokens/colors.json';

class Checkbox extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      value: props.value,
      undetermined: props.undetermined
    };
    this.toggle = this.toggle.bind(this);
  }

  componentWillReceiveProps(props) {
    const { value } = props;
    if (this.state.value != value) this.setState({ value });
  }

  toggle() {
    const { onChange, id, disabled } = this.props;
    if (!disabled) {
      let value = !this.state.value;
      this.setState({ value, undetermined: false });
      if (onChange) onChange(value, id);
    }
  }

  handleKeyDown(e) {
    if (e.keyCode == 13 || e.keyCode == 32) {
      e.preventDefault();
      this.toggle();
    }
  }

  render() {
    const { value, undetermined } = this.state;
    const {
      classes,
      label,
      right,
      disabled,
      textOverflow,
      id,
      className,
      style,
      alignLeft,
      trk,
      testId
    } = this.props;
    let iconClass = undetermined
      ? ` ${classes.undetermined}`
      : value
      ? ` ${classes.active}`
      : '';
    const iconColor = disabled ? colors.icon.brand.disabled : colors.icon.inverse.default;
    return (
      <div
        id={id}
        tabIndex={disabled ? -1 : 0}
        onKeyDown={(e) => this.handleKeyDown(e)}
        className={`${classes.cont}${iconClass}${
          disabled ? ` ${classes.disabled}` : ''
        }${className ? ` ${className}` : ''}`}
        onClick={this.toggle}
        style={style}
        data-testid={testId}
        {...(testId && {
          'data-value': value ? 1 : 0
        })}
      >
        <div className={classes.checkWrap}>
          <div className={classes.check} id={trk}>
            <div className={classes.box} />
            {undetermined && (
              <Icon iconName="minus" width={16} height={16} colors={[iconColor]} className={classes.icon} />
            )}
            {!undetermined && value &&
              <Icon iconName="check" width={16} height={16} colors={[iconColor]} className={classes.icon} />
            }
          </div>
        </div>
        {!!(label || right) && (
          <div className={classes.labelWrap}>
            {label && (
              <Text
                tag="label"
                className={`${classes.label}${
                  textOverflow ? ` ${classes.overflow}` : ''
                }${
                  alignLeft ? ` ${classes.alignLeft}` : ''
                }`}
              >
                {label}
              </Text>
            )}
            {right && (
              <Text tag="label" mid className={classes.right}>
                {right}
              </Text>
            )}
          </div>
        )}
      </div>
    );
  }
}

Checkbox.propTypes = {
  classes: PropTypes.object.isRequired,
  /** Default value for the checkbox. */
  value: PropTypes.bool,
  /** Used for partial check or when is unable to retrieve the value as absolute. */
  undetermined: PropTypes.bool,
  /** The function to call when the value of the checkbox changes. */
  onChange: PropTypes.func,
  /** Disable the checkbox. */
  disabled: PropTypes.bool,
  /** Text to display next to the checkbox.  */
  label: PropTypes.node,
  /** Text to show on the right. */
  right: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /** Use this prop to overflow the text of the label, adding '...' and the end. */
  textOverflow: PropTypes.bool,
  /** Align the content to the left */
  alignLeft: PropTypes.bool,
  /** Id for catching selected element for tracking. */
  trk: PropTypes.string,
  id: PropTypes.string,
  className: PropTypes.string,
  style: PropTypes.object,
  /** The testId property adds the data attribute data-testid to the main element and should be used for testing only. */
  testId: PropTypes.string
};

export default Checkbox;
