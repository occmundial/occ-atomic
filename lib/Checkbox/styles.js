import icons from '../subatomic/icons';
import spacing from '../tokens/spacing.json';
import shadows from '../tokens/shadows.json';
import colors from '../tokens/colors.json';
import borderRadius from '../tokens/borderRadius.json';


const { checkbox, icon } = colors;

export default {
  cont: {
    paddingTop: spacing['size-2'],
    paddingBottom: spacing['size-2'],
    boxSizing: 'border-box',
    display: 'flex',
    alignItems: 'start',
    cursor: 'pointer',
    outline: '0',
    '&:after': {
      content: '""',
      display: 'table',
      clear: 'both'
    },
    '&:focus $check:before': {
      boxShadow: shadows['focus-bright-blue'],
      borderWidth: 1
    },
    '&$active, &$undetermined': {
      '&:hover $check:before, &:active $check:before': {
        background: checkbox['selected']['bg']['hover']
      }
    },
    '&:not($active), &:not($undetermined)': {
      '&:hover $check:before, &:active $check:before': {
        borderColor: checkbox['unselected']['border']['hover']
      },
      '&:active $check:before': {
        borderWidth: 2
      }
    }
  },
  check: {
    width: spacing['size-5'],
    height: spacing['size-5'],
    position: 'relative',
    '&:before': {
      content: '""',
      boxSizing: 'border-box',
      width: 20,
      height: 20,
      borderRadius: borderRadius['br-xs'],
      position: 'absolute',
      top: '50%',
      left: '50%',
      transform: 'translate(-50%, -50%)',
      border: `1px solid ${checkbox['unselected']['border']['default']}`,
      background: checkbox['unselected']['bg']['default'],
      transition: '0.3s all'
    },
    '&:after': {
      content: '""',
      boxSizing: 'border-box',
      width: 20,
      height: 20,
      marginTop: -1,
      position: 'absolute',
      top: '50%',
      left: '50%',
      transform: 'translate(-50%, -50%)',
      transition: '0.3s all',
      background: 'transparent'
    }
  },
  active: {
    '& $check:after': {
      background: icons.base(icons.check.icon([icon['inverse']['default']]))
    },
    '& $check:before': {
      borderColor: checkbox['selected']['border']['default'],
      background: checkbox['selected']['bg']['default']
    }
  },
  undetermined: {
    '& $check:before': {
      borderColor: checkbox['selected']['border']['default'],
      background: checkbox['selected']['bg']['default']
    },
    '& $check:after': {
      background: icons.base(icons.minus.icon([icon['inverse']['default']])),
      marginTop: 0
    }
  },
  disabled: {
    pointerEvents: 'none',
    '&$active, &$undetermined': {
      '& $check:before': {
        borderColor: checkbox['selected']['border']['default'],
        background: checkbox['selected']['bg']['disabled']
      }
    },
    '&$undetermined $check:after': {
      background: icons.base(icons.minus.icon([icon['brand']['disabled']]))
    },
    '&$active $check:after': {
      background: icons.base(icons.check.icon([icon['brand']['disabled']]))
    },
    '&:not($active), &:not($undetermined)': {
      '& $check:before': {
        borderColor: checkbox['unselected']['border']['default'],
        background: checkbox['unselected']['bg']['disabled']
      }
    }
  },
  label: {
    marginLeft: spacing['size-2'],
    cursor: 'pointer',
    float: 'left',
    flex: '1'
  },
  right: {
    marginLeft: spacing['size-2'],
    float: 'right',
    pointerEvents: 'none'
  },
  overflow: {
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap'
  },
  alignLeft: {
    flex: 'none'
  }
};
