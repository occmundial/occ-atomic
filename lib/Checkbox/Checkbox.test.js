import React from 'react';
import { mount } from 'enzyme';
import { JssProvider, SheetsRegistry } from 'react-jss';
import classnames from 'classnames';
import Checkbox from './Checkbox';

let sheets, classes;

describe('Checkbox', () => {

    beforeEach(() => sheets = new SheetsRegistry);
    afterEach(() => sheets = null);

    it('matches the snapshot', () => {
        const wrapper = mount(
            <Checkbox />
        );
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('has the right state', () => {
        let wrapper = mount(
            <JssProvider registry={sheets}>
                <Checkbox />
            </JssProvider>
        );
        classes = sheets.registry[0].classes;
        expect(wrapper.find('div').first().prop('className')).toEqual(classes.cont);
        wrapper = mount(
            <JssProvider registry={sheets}>
                <Checkbox value />
            </JssProvider>
        );
        expect(wrapper.find('div').first().prop('className')).toEqual(classnames(classes.cont, classes.active));
    });

    it('appears undetermined', () => {
        const wrapper = mount(
            <JssProvider registry={sheets}>
                <Checkbox undetermined />
            </JssProvider>
        );
        classes = sheets.registry[0].classes;
        expect(wrapper.find('div').first().prop('className')).toEqual(classnames(classes.cont, classes.undetermined));
    });

    it('is disabled', () => {
        const wrapper = mount(
            <JssProvider registry={sheets}>
                <Checkbox disabled />
            </JssProvider>
        );
        classes = sheets.registry[0].classes;
        expect(wrapper.find('div').first().prop('className')).toEqual(classnames(classes.cont, classes.disabled));
    });

    it('calls onChange function', () => {
        const onChange = jest.fn();
        const wrapper = mount(
            <Checkbox onChange={onChange} />
        );
        wrapper.simulate('click');
        expect(onChange.mock.calls.length).toBe(1);
    });

    it('shows the label', () => {
        const wrapper = mount(
            <JssProvider registry={sheets}>
                <Checkbox label="Label content" right="Right" />
            </JssProvider>
        );
        classes = sheets.registry[1].classes;
        expect(wrapper.find(`.${classes.label}`).first().text()).toEqual('Label content');
    });

});
