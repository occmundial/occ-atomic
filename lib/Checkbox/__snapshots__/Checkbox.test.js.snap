// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox matches the snapshot 1`] = `ShallowWrapper {}`;

exports[`Checkbox styles matches the snapshot 1`] = `
Object {
  "active": Object {
    "& $check:after": Object {
      "background": "url(data:image/svg+xml;base64,PHN2ZyB2ZXJzaW9uPSIxLjEiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgeG1sbnM6eGxpbms9Imh0dHA6Ly93d3cudzMub3JnLzE5OTkveGxpbmsiIHg9IjBweCIgeT0iMHB4IgoJCQkgdmlld0JveD0iMCAwIDI0IDI0IiBzdHlsZT0iZW5hYmxlLWJhY2tncm91bmQ6bmV3IDAgMCAyNCAyNDsiIHhtbDpzcGFjZT0icHJlc2VydmUiPgoJCTxwYXRoIGZpbGw9IiNmZmYiIGQ9Ik0xOC4zLDcuM2wxLjQsMS40bC04LjcsOWMtMC40LDAuNC0xLDAuNC0xLjQsMC4xbC00LjMtMy42bDEuMy0xLjVsMy42LDNDMTAuMiwxNS42LDE4LjMsNy4zLDE4LjMsNy4zeiIvPgoJCTwvc3ZnPg==)",
    },
    "& $check:before": Object {
      "background": "#0059CD",
      "borderColor": "rgba(255,255,255,0.2)",
    },
  },
  "alignLeft": Object {
    "flex": "none",
  },
  "check": Object {
    "&:after": Object {
      "background": "transparent",
      "boxSizing": "border-box",
      "content": "\\"\\"",
      "height": 20,
      "left": "50%",
      "marginTop": -1,
      "position": "absolute",
      "top": "50%",
      "transform": "translate(-50%, -50%)",
      "transition": "0.3s all",
      "width": 20,
    },
    "&:before": Object {
      "background": "#fff",
      "border": "1px solid #D3D4DC",
      "borderRadius": "4px",
      "boxSizing": "border-box",
      "content": "\\"\\"",
      "height": 20,
      "left": "50%",
      "position": "absolute",
      "top": "50%",
      "transform": "translate(-50%, -50%)",
      "transition": "0.3s all",
      "width": 20,
    },
    "height": "24px",
    "position": "relative",
    "width": "24px",
  },
  "cont": Object {
    "&$active, &$undetermined": Object {
      "&:hover $check:before, &:active $check:before": Object {
        "background": "#083CAE",
      },
    },
    "&:after": Object {
      "clear": "both",
      "content": "\\"\\"",
      "display": "table",
    },
    "&:focus $check:before": Object {
      "borderWidth": 1,
      "boxShadow": "0 0 0 8px rgba(0,110,255,0.3)",
    },
    "&:not($active), &:not($undetermined)": Object {
      "&:active $check:before": Object {
        "borderWidth": 2,
      },
      "&:hover $check:before, &:active $check:before": Object {
        "borderColor": "#6C6F89",
      },
    },
    "alignItems": "start",
    "boxSizing": "border-box",
    "cursor": "pointer",
    "display": "flex",
    "outline": "0",
    "paddingBottom": "8px",
    "paddingTop": "8px",
  },
  "disabled": Object {
    "&$active $check:after": Object {
      "background": "url(data:image/svg+xml;base64,PHN2ZyB2ZXJzaW9uPSIxLjEiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgeG1sbnM6eGxpbms9Imh0dHA6Ly93d3cudzMub3JnLzE5OTkveGxpbmsiIHg9IjBweCIgeT0iMHB4IgoJCQkgdmlld0JveD0iMCAwIDI0IDI0IiBzdHlsZT0iZW5hYmxlLWJhY2tncm91bmQ6bmV3IDAgMCAyNCAyNDsiIHhtbDpzcGFjZT0icHJlc2VydmUiPgoJCTxwYXRoIGZpbGw9IiM4REE1REEiIGQ9Ik0xOC4zLDcuM2wxLjQsMS40bC04LjcsOWMtMC40LDAuNC0xLDAuNC0xLjQsMC4xbC00LjMtMy42bDEuMy0xLjVsMy42LDNDMTAuMiwxNS42LDE4LjMsNy4zLDE4LjMsNy4zeiIvPgoJCTwvc3ZnPg==)",
    },
    "&$active, &$undetermined": Object {
      "& $check:before": Object {
        "background": "hsl(221 91.2% 35.7% / 0.1)",
        "borderColor": "rgba(255,255,255,0.2)",
      },
    },
    "&$undetermined $check:after": Object {
      "background": "url(data:image/svg+xml;base64,PHN2ZyB2ZXJzaW9uPSIxLjEiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgeG1sbnM6eGxpbms9Imh0dHA6Ly93d3cudzMub3JnLzE5OTkveGxpbmsiIHg9IjBweCIgeT0iMHB4IgogICAgICAgICAgICB2aWV3Qm94PSIwIDAgMjQgMjQiIHN0eWxlPSJlbmFibGUtYmFja2dyb3VuZDpuZXcgMCAwIDI0IDI0OyIgeG1sOnNwYWNlPSJwcmVzZXJ2ZSI+CiAgICAgICAgPHBvbHlnb24gcG9pbnRzPSI1LjMsMTMuMyA1LjMsMTAuNyAxOC43LDEwLjcgMTguNywxMy4zICIgZmlsbD0iIzhEQTVEQSIgLz4KICAgICAgICA8L3N2Zz4=)",
    },
    "&:not($active), &:not($undetermined)": Object {
      "& $check:before": Object {
        "background": "#EDEDF1",
        "borderColor": "#D3D4DC",
      },
    },
    "pointerEvents": "none",
  },
  "label": Object {
    "cursor": "pointer",
    "flex": "1",
    "float": "left",
    "marginLeft": "8px",
  },
  "overflow": Object {
    "overflow": "hidden",
    "textOverflow": "ellipsis",
    "whiteSpace": "nowrap",
  },
  "right": Object {
    "float": "right",
    "marginLeft": "8px",
    "pointerEvents": "none",
  },
  "undetermined": Object {
    "& $check:after": Object {
      "background": "url(data:image/svg+xml;base64,PHN2ZyB2ZXJzaW9uPSIxLjEiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgeG1sbnM6eGxpbms9Imh0dHA6Ly93d3cudzMub3JnLzE5OTkveGxpbmsiIHg9IjBweCIgeT0iMHB4IgogICAgICAgICAgICB2aWV3Qm94PSIwIDAgMjQgMjQiIHN0eWxlPSJlbmFibGUtYmFja2dyb3VuZDpuZXcgMCAwIDI0IDI0OyIgeG1sOnNwYWNlPSJwcmVzZXJ2ZSI+CiAgICAgICAgPHBvbHlnb24gcG9pbnRzPSI1LjMsMTMuMyA1LjMsMTAuNyAxOC43LDEwLjcgMTguNywxMy4zICIgZmlsbD0iI2ZmZiIgLz4KICAgICAgICA8L3N2Zz4=)",
      "marginTop": 0,
    },
    "& $check:before": Object {
      "background": "#0059CD",
      "borderColor": "rgba(255,255,255,0.2)",
    },
  },
}
`;

exports[`CheckboxJSS matches the snapshot 1`] = `ShallowWrapper {}`;
